File: Components/Exceptions/LocalizedText.vb
Imports DotNetNuke.Services.Localization
Imports System.Runtime.Serialization

Namespace DotNetNuke.Modules.Reports.Exceptions
    <Serializable()> _
    Public Class LocalizedText
        Implements ISerializable

#Region " Constants "

        Private Const SER_KEY_ResourceKey As String = "ResourceKey"
        Private Const SER_KEY_ResourceFile As String = "ResourceFile"
        Private Const SER_KEY_FormatArgs As String = "FormatArg"
        Private Const SER_KEY_FormatArgsCount As String = "FormatArgsCount"
        Private Const RESX_ModuleCommon As String = "~/DesktopModules/Reports/App_LocalResources/SharedResources.resx"


#End Region

        Private _resourceKey As String
        Private _resourceFile As String
        Private _formatArguments As String()

        Public ReadOnly Property ResourceKey() As String
            Get
                Return _resourceKey
            End Get
        End Property

        Public ReadOnly Property ResourceFile() As String
            Get
                Return _resourceFile
            End Get
        End Property

        Public ReadOnly Property FormatArguments() As String()
            Get
                Return _formatArguments
            End Get
        End Property

        Public Sub New(ByVal resourceKey As String)
            Me.New(resourceKey, RESX_ModuleCommon, New String() {})
        End Sub

        Public Sub New(ByVal resourceKey As String, ByVal formatArguments As String())
            Me.New(resourceKey, RESX_ModuleCommon, formatArguments)
        End Sub

        Public Sub New(ByVal resourceKey As String, ByVal resourceFile As String, ByVal ParamArray formatArguments As String())
            _resourceKey = resourceKey
            _resourceFile = resourceFile
            _formatArguments = formatArguments
        End Sub

        Public Sub New(ByVal info As SerializationInfo, ByVal context As StreamingContext)
            Me._resourceKey = info.GetString(SER_KEY_ResourceKey)
            Me._resourceFile = info.GetString(SER_KEY_ResourceFile)

            Dim count As Integer = info.GetInt32(SER_KEY_FormatArgsCount)
            ReDim _formatArguments(count)
            For i As Integer = 0 To count
                _formatArguments(i) = info.GetString(String.Concat(SER_KEY_FormatArgs, i.ToString))
            Next
        End Sub

        Public Function GetLocalizedText() As String
            Dim localized As String = String.Empty
            If Not String.IsNullOrEmpty(_resourceFile) Then
                localized = Localization.GetString(_resourceKey, _resourceFile)
            Else
                localized = Localization.GetString(_resourceKey)
            End If
            If _formatArguments IsNot Nothing AndAlso _formatArguments.Length > 0 Then
                Return String.Format(localized, _formatArguments)
            Else
                Return localized
            End If
        End Function

        Public Sub GetObjectData(ByVal info As SerializationInfo, ByVal context As StreamingContext) Implements ISerializable.GetObjectData
            info.AddValue(SER_KEY_ResourceKey, Me._resourceKey)
            info.AddValue(SER_KEY_ResourceFile, Me._resourceFile)
            info.AddValue(SER_KEY_FormatArgsCount, Me._formatArguments.Length)
            For i As Integer = 0 To Me._formatArguments.Length
                info.AddValue(String.Concat(SER_KEY_FormatArgs, i.ToString()), Me._formatArguments(i))
            Next
        End Sub
    End Class
End Namespace

File: Components/Exceptions/LocalizedText.cs
#region Copyright
// 
// DotNetNuke® - http://www.dotnetnuke.com
// Copyright (c) 2002-2018
// by DotNetNuke Corporation
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated 
// documentation files (the "Software"), to deal in the Software without restriction, including without limitation 
// the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and 
// to permit persons to whom the Software is furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in all copies or substantial portions 
// of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED 
// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL 
// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF 
// CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
// DEALINGS IN THE SOFTWARE.
//
#endregion


namespace DotNetNuke.Modules.Reports.Exceptions
{
    using System;
    using System.Runtime.Serialization;
    using DotNetNuke.Services.Localization;

    [Serializable]
    public class LocalizedText : ISerializable
    {
        public LocalizedText(string resourceKey) : this(resourceKey, RESX_ModuleCommon)
        { }

        public LocalizedText(string resourceKey, string[] formatArguments) : this(
            resourceKey, RESX_ModuleCommon, formatArguments)
        { }

        public LocalizedText(string resourceKey, string resourceFile, params string[] formatArguments)
        {
            this.ResourceKey = resourceKey;
            this.ResourceFile = resourceFile;
            this.FormatArguments = formatArguments;
        }

        public LocalizedText(SerializationInfo info, StreamingContext context)
        {
            this.ResourceKey = info.GetString(SER_KEY_ResourceKey);
            this.ResourceFile = info.GetString(SER_KEY_ResourceFile);

            var count = info.GetInt32(SER_KEY_FormatArgsCount);
            this.FormatArguments = new string[count + 1];
            for (var i = 0; i <= count; i++)
            {
                this.FormatArguments[i] = info.GetString(string.Concat(SER_KEY_FormatArgs, i.ToString()));
            }
        }

        public string ResourceKey { get; }

        public string ResourceFile { get; }

        public string[] FormatArguments { get; }

        public void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            info.AddValue(SER_KEY_ResourceKey, this.ResourceKey);
            info.AddValue(SER_KEY_ResourceFile, this.ResourceFile);
            info.AddValue(SER_KEY_FormatArgsCount, this.FormatArguments.Length);
            for (var i = 0; i <= this.FormatArguments.Length; i++)
            {
                info.AddValue(string.Concat(SER_KEY_FormatArgs, i.ToString()), this.FormatArguments[i]);
            }
        }

        public string GetLocalizedText()
        {
            var localized = string.Empty;
            if (!string.IsNullOrEmpty(this.ResourceFile))
            {
                localized = Localization.GetString(this.ResourceKey, this.ResourceFile);
            }
            else
            {
                localized = Localization.GetString(this.ResourceKey);
            }
            if (this.FormatArguments != null && this.FormatArguments.Length > 0)
            {
                return string.Format(localized, this.FormatArguments);
            }
            return localized;
        }

        #region  Constants

        private const string SER_KEY_ResourceKey = "ResourceKey";
        private const string SER_KEY_ResourceFile = "ResourceFile";
        private const string SER_KEY_FormatArgs = "FormatArg";
        private const string SER_KEY_FormatArgsCount = "FormatArgsCount";
        private const string RESX_ModuleCommon = "~/DesktopModules/Reports/App_LocalResources/SharedResources.resx";

        #endregion
    }
}